ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/BlockValidator.java

```diff
@@ -13,187 +13,91 @@
-  private final GossipValidationHelper gossipValidationHelper;
-  private final Set<SlotAndProposer> receivedValidBlockInfoSet =
-      LimitedSet.createSynchronized(VALID_BLOCK_SET_SIZE);
-
-  public BlockValidator(
-      final Spec spec,
-      final RecentChainData recentChainData,
-      final GossipValidationHelper gossipValidationHelper) {
-    this.spec = spec;
-    this.recentChainData = recentChainData;
-    this.gossipValidationHelper = gossipValidationHelper;
-  }
-
-  public SafeFuture<InternalValidationResult> validate(final SignedBeaconBlock block) {
-
-    if (gossipValidationHelper.isSlotFinalized(block.getSlot())
-        || !blockIsFirstBlockWithValidSignatureForSlot(block)) {
-      LOG.trace(
-          "BlockValidator: Block is either too old or is not the first block with valid signature for "
-              + "its slot. It will be dropped");
-      return completedFuture(InternalValidationResult.IGNORE);
-    }
 
-    if (gossipValidationHelper.isSlotFromFuture(block.getSlot())) {
-      LOG.trace("BlockValidator: Block is from the future. It will be saved for future processing");
-      return completedFuture(InternalValidationResult.SAVE_FOR_FUTURE);
-    }
-
-    if (gossipValidationHelper.isBlockAvailable(block.getRoot())) {
-      LOG.trace("Block is already imported");
-      return completedFuture(InternalValidationResult.IGNORE);
-    }
+  private final BlockGossipValidator blockGossipValidator;
 
-    if (!gossipValidationHelper.isBlockAvailable(block.getParentRoot())) {
-      LOG.trace("Block parent is not available. It will be saved for future processing");
-      return completedFuture(InternalValidationResult.SAVE_FOR_FUTURE);
-    }
-
-    if (!currentFinalizedCheckpointIsAncestorOfBlock(block)) {
-      return completedFuture(reject("Block does not descend from finalized checkpoint"));
-    }
+  public BlockValidator(final BlockGossipValidator blockGossipValidator) {
+    this.blockGossipValidator = blockGossipValidator;
+  }
 
-    final Optional<UInt64> maybeParentBlockSlot =
-        gossipValidationHelper.getSlotForBlockRoot(block.getParentRoot());
-    if (maybeParentBlockSlot.isEmpty()) {
-      LOG.trace(
-          "BlockValidator: Parent block does not exist. It will be saved for future processing");
-      return completedFuture(InternalValidationResult.SAVE_FOR_FUTURE);
-    }
-    final UInt64 parentBlockSlot = maybeParentBlockSlot.get();
+  public SafeFuture<InternalValidationResult> validateGossip(final SignedBeaconBlock block) {
+    return blockGossipValidator.validate(block, false);
+  }
 
-    if (parentBlockSlot.isGreaterThanOrEqualTo(block.getSlot())) {
-      return completedFuture(reject("Parent block is after child block."));
+  public SafeFuture<BroadcastValidationResult> validateBroadcast(
+      final SignedBeaconBlock block,
+      final BroadcastValidationLevel broadcastValidationLevel,
+      final SafeFuture<BlockImportResult> consensusValidationResult) {
```

Comment:
Are you sure on this interface? In case of GOSSIP validation we don't need this SafeFuture but still need to provide it.

Reply:
ye, you're actually true that we could avoid tracking consensus if validation level is GOSSIP, but the internal validation logic will percolate outside the `BlockValidator` class, with more `if`s in `BlockManager`, and with imperceptible perf benefits. So I'm not sure..

Reply:
IMHO the way we have the rules "pipeline" in here with each step next to each other helps readability. Even if the consensus validations result is not needed for one case, at least things are contained here.

I think we should keep it this way :)

Reply:
The question is whether we have consensusValidationResult when we call this method with `GOSSIP` validation level. If yes - this interface is ok. If we need to pass some SafeFuture stub instead when we have GOSSIP validation, it will stink a bit.